services/server/server.go

```diff
@@ -493,6 +496,70 @@ func (s *Server) ChangeSettings(ctx context.Context, req *serverpb.ChangeSetting
 	return res, nil
 }
 
+// SetSSHKey uploads public key.
+func (s *Server) SetSSHKey(ctx context.Context, in *serverpb.SetSSHKeyRequest) (*serverpb.SetSSHKeyResponse, error) {
+	//TODO: validate ssh key
+	var settings *models.Settings
+	err := s.db.InTransaction(func(tx *reform.TX) error {
+		var e error
+		if settings, e = models.GetSettings(tx); e != nil {
+			return e
+		}
+
+		username := "admin"
+		usr, err := user.Lookup(username)
+		if err != nil {
+			return err
+		}
+
+		uid, err := strconv.Atoi(usr.Uid)
+		if err != nil {
+			return err
+		}
+		gid, err := strconv.Atoi(usr.Gid)
+		if err != nil {
+			return err
+		}
+
+		sshDirPath := path.Join(usr.HomeDir, ".ssh")
+		if _, err := os.Stat(sshDirPath); os.IsNotExist(err) {
+			err := os.Mkdir(sshDirPath, os.FileMode(0700))
+			if err != nil {
+				return err
+			}
+			err = os.Chown(sshDirPath, uid, gid)
+			if err != nil {
+				return err
+			}
+		}
+
+		keysPath := path.Join(sshDirPath, "authorized_keys")
+
+		file, err := os.OpenFile(keysPath, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, os.FileMode(0600))
+		if err != nil {
+			return err
+		}
+
+		_, err = file.WriteString(in.SshKey)
```

Comment:
Note that file is never closed there. That's a big problem.

Use `ioutil.WriteFile` to simplify the whole thing.

Reply:
It still wasn't being closed.